File: cpp/64.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int minPathSum(vector<vector<int> > &grid) {
    int m = grid.size();
    int n = grid[0].size();
    vector<int> dp(n, 0);
    for(int i = 0; i < m; i++){
      for(int j = 0; j < n; j++){
        if(!i && !j)
          dp[j] = grid[i][j];
        else
          dp[j] = grid[i][j] + min(i?dp[j]:INT_MAX, j?dp[j-1]:INT_MAX);
      }
      prints(dp);
    }
    return dp[n-1];
  }
};

int main(){
  auto sol = Solution();
  vector<vector<int> > v={{1,2},{1,1}};
  sol.minPathSum(v);
}

File: python/64.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        n = len(grid)
        if not n:
            return 0
        m = len(grid[0])
        for i in xrange(n):
            for j in xrange(m):
                if i and j:
                    grid[i][j] += min(grid[i][j-1], grid[i-1][j])
                elif i:
                    grid[i][j] += grid[i-1][j]
                elif j:
                    grid[i][j] += grid[i][j-1]
        return grid[n-1][m-1]
